File: app/lib/src/time_format.dart
/// Lifetime formatting for the river's countdown rings and pills.
///
/// Urgency is *relative* to a post's original lifetime (the design's ring
/// fraction is `remaining / originalLifetime`): the amber "draining"
/// treatment kicks in over the final fifth, critical over the final
/// twentieth — a 6h post and a 1-year post both drain over their own last
/// stretch rather than at a fixed wall-clock threshold.
library;

/// How urgently a post's remaining lifetime should read.
enum UrgencyTier {
  /// Permanent — no lifetime at all.
  settled,
  normal,

  /// The final fifth of the post's lifetime: the amber treatment.
  warm,

  /// The final twentieth (or already expired).
  critical,
}

const double _warmFraction = 0.20;
const double _criticalFraction = 0.05;

typedef LifetimeState = ({String label, UrgencyTier tier, double fraction});

/// Remaining-lifetime label, urgency tier, and ring fraction (0..1) for an
/// ebbing post. `expiresAt <= now` renders as "ebbing away" at fraction 0.
LifetimeState lifetimeState({
  required int now,
  required int createdAt,
  required int expiresAt,
}) {
  final remaining = expiresAt - now;
  if (remaining <= 0) {
    return (label: 'ebbing away', tier: UrgencyTier.critical, fraction: 0);
  }
  final original = expiresAt - createdAt;
  final fraction = original > 0 ? (remaining / original).clamp(0.0, 1.0) : 0.0;
  final tier = fraction <= _criticalFraction
      ? UrgencyTier.critical
      : fraction <= _warmFraction
      ? UrgencyTier.warm
      : UrgencyTier.normal;
  return (label: formatRemaining(remaining), tier: tier, fraction: fraction);
}

/// Formats a remaining duration in seconds: "3d", "34h", "4h12m", "58m",
/// "<1m".
String formatRemaining(int remaining) {
  if (remaining >= 48 * 3600) return '${remaining ~/ (24 * 3600)}d';
  if (remaining >= 10 * 3600) return '${remaining ~/ 3600}h';
  if (remaining >= 3600) {
    final minutes = (remaining % 3600) ~/ 60;
    return '${remaining ~/ 3600}h${minutes.toString().padLeft(2, '0')}m';
  }
  if (remaining >= 60) return '${remaining ~/ 60}m';
  return '<1m';
}

/// Compact age for author rows: "now", "5m", "2h", "3d".
String formatAge(int now, int createdAt) {
  final age = now - createdAt;
  if (age < 60) return 'now';
  if (age < 3600) return '${age ~/ 60}m';
  if (age < 48 * 3600) return '${age ~/ 3600}h';
  return '${age ~/ (24 * 3600)}d';
}

int nowUnixSecs() => DateTime.now().millisecondsSinceEpoch ~/ 1000;

File: src/time_format.rs
//! Lifetime formatting for the river's countdown pills.

/// How urgently a post's remaining lifetime should read.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) enum UrgencyTier {
    /// Permanent — no lifetime at all.
    Settled,
    Normal,
    /// Less than six hours left: the warm/amber treatment.
    Warm,
    /// Less than one hour left.
    Critical,
}

const WARM_THRESHOLD_SECS: u64 = 6 * 3600;
const CRITICAL_THRESHOLD_SECS: u64 = 3600;

/// Formats the remaining lifetime of a post ("34h", "4h12m", "58m", "<1m")
/// with its urgency tier. `expires_at <= now` renders as "ebbing away".
pub(crate) fn format_remaining(now: u64, expires_at: u64) -> (String, UrgencyTier) {
    let Some(remaining) = expires_at.checked_sub(now).filter(|left| *left > 0) else {
        return ("ebbing away".to_owned(), UrgencyTier::Critical);
    };

    let tier = if remaining < CRITICAL_THRESHOLD_SECS {
        UrgencyTier::Critical
    } else if remaining < WARM_THRESHOLD_SECS {
        UrgencyTier::Warm
    } else {
        UrgencyTier::Normal
    };

    let label = if remaining >= 48 * 3600 {
        format!("{}d", remaining / (24 * 3600))
    } else if remaining >= 10 * 3600 {
        format!("{}h", remaining / 3600)
    } else if remaining >= 3600 {
        format!("{}h{:02}m", remaining / 3600, (remaining % 3600) / 60)
    } else if remaining >= 60 {
        format!("{}m", remaining / 60)
    } else {
        "<1m".to_owned()
    };

    (label, tier)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn remaining_lifetimes_format_with_urgency_tiers() {
        assert_eq!(
            format_remaining(0, 3 * 24 * 3600),
            ("3d".to_owned(), UrgencyTier::Normal)
        );
        assert_eq!(
            format_remaining(0, 34 * 3600),
            ("34h".to_owned(), UrgencyTier::Normal)
        );
        assert_eq!(
            format_remaining(0, 4 * 3600 + 12 * 60),
            ("4h12m".to_owned(), UrgencyTier::Warm)
        );
        assert_eq!(
            format_remaining(0, 58 * 60),
            ("58m".to_owned(), UrgencyTier::Critical)
        );
        assert_eq!(
            format_remaining(0, 30),
            ("<1m".to_owned(), UrgencyTier::Critical)
        );
        assert_eq!(
            format_remaining(100, 100),
            ("ebbing away".to_owned(), UrgencyTier::Critical)
        );
    }
}
